design-proposals/bound-token-for-vnc.md

```diff
@@ -0,0 +1,110 @@
+# Overview
+
+Kubernetes API server can create 
+[tokens](https://github.com/kubernetes/enhancements/tree/master/keps/sig-auth/1205-bound-service-account-tokens)
+bound to a service account. They are mainly used in services running inside the cluster to authenticate
+between each other. 
+
+This document describes how to create a token that can be used to authenticate with the API server
+to only access VNC endpoint of a singe VM.
+
+## Motivation
+
+An administrator may want to give VNC access to a client application without giving it permissions to access
+anything else in the cluster. The VNC client could be pointed to VMI `/vnc` subresource and use
+the token to authenticate. 
+
+## Goals
+
+- A temporary token can be created to access VNC.
+- It can only be used with one specific VM.
+- It is valid only for a specific duration.
```

Comment:
Will there be a way to specify no expiration?

For example, if I am an admin and I create VMs for users to use as desktops, I would not want to keep recreating the tokens.
Or if a user needs to run other applications inside the VM for their normal job. Such as they have a linux desktop on their laptop for daily use but sometimes need to run an app inside windows.

I suspect a method to allow the user to renew their token would be a potential security issue.

Reply:
> For example, if I am an admin and I create VMs for users to use as desktops, I would not want to keep recreating the tokens.

VNC over websockets in KubeVirt is explicitly not recommended for desktop usage. Just to make that clear, while it is great if you can use it that way for your use cases. :) 

The vga adapter has for instance a very small hardcoded amount of memory and since this  goes over the apiserver it may have unintended side effects if this is used by a bigger group of engineers as real desktop over this VNC session.

> For example, if I am an admin and I create VMs for users to use as desktops, I would not want to keep recreating the tokens.

I am not sure if this proposal is covering that use-case. @akrejcir ? I would expect that for permanent access, you can already right now create a ServiceAccount, Role and RoleBinding and maintain it like all other RBAC permissions for your cluster. The VNC token idea may sound appealing initially because you don't have to create these objects, bit in practice it probably means for admins more hassle to manage permanent RBAC access of some features without direct yaml applying.

What do you think?

Reply:
Yes, it does not cover the use-case of permanent tokens. They would need a different implementation.
I think for now we can keep this design only for temporary tokens, and permanent access can be created using RBAC, as you've mentioned.